### PR TITLE
Optimize away some physical equality

### PR DESCRIPTION
**Pull request by chambart**
Physical equality on structurally different values is always false. This patch allows to handle some cases. The need arise when evaluating some functions that use it to avoid some computation. For instance in `Set.add` test if the added value is already present to avoid rebuilding the same set.

``` OCaml
    let rec add x = function
        Empty -> Node(Empty, x, Empty, 1)
      | Node(l, v, r, _) as t ->
          let c = Ord.compare x v in
          if c = 0 then t else
          if c < 0 then
            let ll = add x l in
            if l == ll then t else bal ll v r
          else
            let rr = add x r in
            if r == rr then t else bal l v rr
```

This change allows some of those uses to be eliminated.

### CODE REVIEW

File: middle_end/simplify_primitives.ml
Comment (mshinwell) [inline middle_end/simplify_primitives.ml]: Please make both of these matches exhaustive.

File: middle_end/simplify_primitives.ml
Comment (mshinwell) [inline middle_end/simplify_primitives.ml]: Indentation needs fixing for these two stanzas (also on line 69 too).
I would also recommend following the GNU convention of putting the `||` at the start of each line of a multi-line conditional; I think it's easier to follow.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
@chambart Ping

File: testsuite/tests/asmcomp/is_static_flambda.ml
Comment (mshinwell) [inline testsuite/tests/asmcomp/is_static_flambda.ml:129]: These tests might as well check that `eq` is as expected, no?

**Comment by mshinwell** [comment]
I made one comment, but OK otherwise.  We need to wait for @lpw25 to review too.

**Comment by mshinwell** [comment]
Ah, this should have a Changes entry too, I think.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
@chambart Please add a Changes entry and merge to trunk, thanks.

**Comment by alainfrisch** [comment]
Ping @chambart

File: testsuite/tests/asmcomp/is_static_flambda.ml
Comment (mshinwell) [reply testsuite/tests/asmcomp/is_static_flambda.ml:129]: Actually I think this is fine as-is.

**Comment by mshinwell** [comment]
I've added the Changes entry and merged with current trunk.  I will merge into trunk assuming CI passes.

**Pull request merged**
